### ISSUE TITLE
Slow editor/project startup times due to shader cache being rebuilt on every launch

### ISSUE THREAD

**Issue by XxlemondevxX** [godotengine/godot#74292]
### Godot version

4.0

### System information

Windows 10.

### Issue description

Godot 3.x worked completely fine and well. Now, whenever I launch a project in the editor it freezes for 20-60 seconds at a time. And then when I want to test my game I have to wait another 20-60 seconds before the game even loads.

### Steps to reproduce

No clue if reproduceable.

### Minimal reproduction project

[Freezie Freeze.zip](https://github.com/godotengine/godot/files/10881830/Freezie.Freeze.zip)
Idk if this will reproduce the same results.

**Comment by AThousandShips** [comment]
Does it randomly freeze or just on start?

**Comment by XxlemondevxX** [comment]
What do you mean randomly?

> Does it randomly freeze or just on start?

**Comment by AThousandShips** [comment]
Does it freeze on start or now and then?

**Comment by XxlemondevxX** [comment]
just on start when opening a game in editor and freezes when i try to run a game through the editor

**Comment by Calinou** [comment]
What CPU and graphics card model do you have?

A slower startup is expected on the initial shaders to allow for shaders to cache. This shouldn't occur on subsequent starts of the editor or a given project though, as Godot will load the cached shaders instead of regenerating them. (Note that updating your graphics drivers will invalidate this cache.)

**Comment by XxlemondevxX** [comment]
I'm running a surface pro with 16gb ddr4, intel i5-1135G7, no GPU because it's a surface pro, no USB peripherals connected.

**Comment by hudwave** [comment]
I also have this issue on a Surface Pro 8 (i7-1185G7 / Iris Xe) with no USB peripherals connected.

4.0 RC6 works fine, loading and running a project is near instantaneous.
4.0 Stable and 4.0.1 RC take ~15s to open then another 15s to run even a newly created blank project.

**Comment by hudwave** [comment]
I've compared the verbose logging output for both 4.0 RC6 and 4.0.1 RC1 and in 4.0.1 RC1 it is much slower to output each of these shader lines.

```
Shader 'CanvasSdfShaderRD' SHA256: 18e5325b1cb73d660b02bf6b71b1ab9d1540e8f0f8808aa0edd78f2e499ef3d9
Shader 'SkeletonShaderRD' SHA256: da10c25e6b40370f600a6fe4fc4aa8b101a931c4fd27cd6087fd0670e4b823db
Shader 'ParticlesShaderRD' SHA256: 3914f47f6432045bc35401a046414265b2389d4a834d2a8e09ad6fca6012cf3f
Shader 'ParticlesCopyShaderRD' SHA256: 83fbf208a122bbbef1288c5ea0b6af69e19e1638b3736a2b2925f6428a1ac088
Shader 'CanvasShaderRD' SHA256: 421de506aac11fb4cbfc19f3bd4e079d2409e856f3bf056bf6215a2392d21331
Shader 'CanvasOcclusionShaderRD' SHA256: a242c5b142c30a3d2e1bad27b01777f7e3f67217cffe572ffc1e97a9f51b8a03
Shader 'ClusterRenderShaderRD' SHA256: c849bc8b5b076a09d76e586dd48e9b4b9d8a0a565463b76d6be55e8b49bcebef
Shader 'ClusterStoreShaderRD' SHA256: 54799cd704d145305bd14699b13a8a1365f75cecce03b11627fe9007f9f728bc
Shader 'ClusterDebugShaderRD' SHA256: 03c5919e1ed3a322c1a47bfd7ae256b29d7e082b5a2dda94330a94608d3d1e61
Shader 'SceneForwardClusteredShaderRD' SHA256: b7c87b5a8c98bba50b9b5ce0dc0736cdba606f950f8d941ecde9220bfafe4722
Shader 'ResolveShaderRD' SHA256: 10c5751d2564a01e9d2ca8822ecef132a07884f4a1b9e5991cb03168fca59e71
Shader 'TaaResolveShaderRD' SHA256: ab879bcb6149354f3d87c9f577282c457ea970ddee811d2f6497e336801e0c0e
Shader 'SsEffectsDownsampleShaderRD' SHA256: 5d6dcb061bbbb22efdefc75564506217590b380a4ec89c2adffa3d9e313bfd44
Shader 'SsilShaderRD' SHA256: 083bad19c0899f05bf4b0674fa5735572fa9abb1e8847f693ca5abb77466163f
Shader 'SsilImportanceMapShaderRD' SHA256: 7429b3885300d62f70e610a551febcbfe1aae686d445efb6934af560b45d677c
Shader 'SsilBlurShaderRD' SHA256: 135feac37842109c2cd43a0981901f13ca782c24516f63df6f5e9b178e1824c5
Shader 'SsilInterleaveShaderRD' SHA256: 089fd141fb6be32e45315ad1fd64727f1949bc343cedec2f6bf8a2db441eca0d
Shader 'SsaoShaderRD' SHA256: 16e5658e46e33902dc0575a335cce7731d5feb919db2374fb37dea904acd12ed
Shader 'SsaoImportanceMapShaderRD' SHA256: 090099f733a91669531d332638a05075875088be668b355b69f3c6931cdf079a
Shader 'SsaoBlurShaderRD' SHA256: 68c92d449ea303260c0e0a0418afe0f4e40a6baf48f1b594665b2b65ef021374
Shader 'SsaoInterleaveShaderRD' SHA256: 2dda0201b5e92dc7c63244b2d5415798a104697575c52482d77939a3da1daa15
Shader 'ScreenSpaceReflectionScaleShaderRD' SHA256: 1405f8e6cc4d4c464152a2abe63855587ef0abd74d2ab7ab3eae7fbe360c9530
Shader 'ScreenSpaceReflectionShaderRD' SHA256: 86af0f448a6a6e2d91a9d6b5caf0fc30017dd4d470195d83ee91106e504ee0c2
Shader 'ScreenSpaceReflectionFilterShaderRD' SHA256: 89704d3164650a97e984b5fbae8b0801fbde0b6fc144a51817299078562ec93d
Shader 'SubsurfaceScatteringShaderRD' SHA256: e5f679c335a3878411a3e0ab12cdc5af65b1ecf98ac3b2d2ce00574e1d717b20
Shader 'SkyShaderRD' SHA256: 3cb6e70446a19cb15d8c1a3f76d56d77f445979a95921e0458bb3db7e5f09dcf
Shader 'VoxelGiShaderRD' SHA256: 651e2bb98c4802d0327ce34cfa849c05e1adb7c9ebab29b51f017299fe0c7577
Shader 'VoxelGiDebugShaderRD' SHA256: 918a9f970f26a10fc7d935d03b2f9713d59172316b0c05c7558a5f6f24945599
Shader 'SdfgiPreprocessShaderRD' SHA256: 3c1f8ad3385508f69572db072ed29a83133b5cf1f4142b2fb377550fa8ec2029
Shader 'SdfgiDirectLightShaderRD' SHA256: e5c72243380a2de86722135b1d91960fe9ede5044261a625b4661c81b9a0337a
Shader 'SdfgiIntegrateShaderRD' SHA256: 430cc27e91f0f0d8fa6b9024856f0de8d059a62fb2d1fcf26cee63a2a9088c37
Shader 'GiShaderRD' SHA256: 1cadddf25933bc24926c894c64474159dfecdf2c3bb1492e40d3dc94cad5af67
Shader 'SdfgiDebugShaderRD' SHA256: 06fd6d4c79f51224803c71fadc7fc20122c959997f208b146cbd37df4358505b
Shader 'SdfgiDebugProbesShaderRD' SHA256: 361efe0ef94133279326010a090095d367736d9eb07bb7ab7e234fab2824004a
Shader 'VolumetricFogShaderRD' SHA256: 5ae1edbd2c32ff8ae5015bfc34d88c3eeaa60e4c609c3b0701d18810f7123ee1
Shader 'VolumetricFogProcessShaderRD' SHA256: 45fdbf9b7c6a8165a282b8c5f1e7f3113fe9376207afeb065eafcc10edc05ab0
Shader 'BokehDofShaderRD' SHA256: cde9cbcfd70ee5d8aba61c4012ea042971689698b88ea6c8efd82ad42d474731
Shader 'CopyShaderRD' SHA256: 9af4d03981f857cf18b72a3f0cb3aa98e31e9f33d8aa9c4201bf0954ae7ac6cb
Shader 'CopyToFbShaderRD' SHA256: 8a2c1902ba5cba7020e6b75246143132a44aba560cf3528a6a02bc4a7aeea975
Shader 'CubeToDpShaderRD' SHA256: 2f0b6c7c665d79ce2adc794f62f8b91a5f87250ae5e3cbba8f2c3b5d2bda0878
Shader 'CubemapDownsamplerShaderRD' SHA256: 74b98c730726d135f6f5c06ee248bc6f2596831a2a2b892720968ad08ab28041
Shader 'CubemapFilterShaderRD' SHA256: 9ec5a93ad259c9ed09b161ef2cfb34316d274311be6b7be14094db3c857d52d7
Shader 'CubemapRoughnessShaderRD' SHA256: b16ee1e96a7339cb1f6d6b8d4e95fa204fe1f29d1f166500be31bbaa35ce5117
Shader 'SpecularMergeShaderRD' SHA256: 7f4e508297da794df1868eb3f9848961b6cdf89820010d17c939b81c6400db77
Shader 'LuminanceReduceShaderRD' SHA256: 698fc0c4856e16e9d52bd257baad7c4068b7b89f27178461a5c04c9ce628cd83
Shader 'TonemapShaderRD' SHA256: 36eb8eb7ea5d613585de9a49b716893c7bcb64b8fc1e6d4faf4fdc0b792f0281
Shader 'FsrUpscaleShaderRD' SHA256: 891d02dbea2f4d35cb13c9c54b1b74add19b3f741728bae71657a4ded5d7e427
Shader 'RoughnessLimiterShaderRD' SHA256: 97414325c38bf4b2c86978214be9152e018e62944441612baa414b0962047596
Shader 'SortShaderRD' SHA256: 2e5d72dd87cc9df342d0cae5f74eaf04f8659d58523c1ded3cc1567c80164961
Shader 'BlitShaderRD' SHA256: 1f149655c961350de64e4639bfa103f7cff7ef5e0e0812e783d03d169ecbdb5b
```

It's slower for all of them, but specifically it seems to hang for ages just after this line has been added

```
Shader 'GiShaderRD' SHA256: 1cadddf25933bc24926c894c64474159dfecdf2c3bb1492e40d3dc94cad5af67
```

Incidentally I've also noticed that instancing a 3d scene is also slower in 4.0.1 once the game is already running.

**Comment by CatastrophicDino** [comment]
Same issue here, I have a GTX 1650 and an Intel Core i5-10400f with 32gb ram. Not sure if this is important but I have Godot installed on an HDD but Windows on an SSD.

**Comment by Calinou** [comment]
I can't reproduce this on 4.0.stable (Linux, GeForce RTX 4090 with NVIDIA 525.89.02). This is likely Windows-specific.

**Comment by XxlemondevxX** [comment]
Seems to also happen on my pretty high end pc...

**Comment by fractalcounty** [comment]
I've been experiencing similar issues on my Windows 11 PC. It's sporadic- sometimes the editor launches extremely slowly, sometimes it doesn't. I do suspect shaders to be the culprit here.

However, after a particularly bad bout of performance issues, I checked my project directory and found a scene file containing a single Label that was over 100MB in size. Later, I found two more that were over half a gigabyte in size. Before I could open an issue, the project crashed and the files went back to normal.

I suspect the culprit was related to FontVariations, because I haven't ran into the issue since I stopped using them. This may be unrelated, but check your project directory for any unusually massive files whenever you experience these issues during runtime

**Comment by kiprou** [comment]
Slow launch (20–30s) for me too on Godot 4.0.stable, even when loading a tiny project. Once the project loads the editor still feels a bit slow/laggy at times, especially when doing stuff in the FileSystem panel.

I'm using Windows 10 Home x64 (21H2), i7-3770K, GTX970 (457.30), 16GB RAM. Like CatastrophicDino above, my OS lives on an SSD, but Godot lives on a HDD (both fast drives).

**Comment by Calinou** [comment]
> but Godot lives on a HDD (both fast drives).

Can you reproduce this issue if you move Godot to a HDD?

**Comment by adcoding** [comment]
I'm having the same issue with Godot 4.0.2 (Steam). It's super slow both to load the editor and to launch a project.

Windows 11
AMD Ryzen 5 7600X
Radeon RX 6800 16G

**Comment by crestofthebeast** [comment]
Experiencing similar issues.
Project runs perfectly on an M1 mac mini, on a windows pc (specs at bottom) I get a couple seconds of hang before the game boots. I do not experience this issue with Godot 3 LTS on the dodge the creeps demo.

```
Windows 11 Pro
AMD Ryzen 5 3600
GTX 3080
X570 Aorus Elite 
```

**Comment by Calinou** [comment]
If you start the editor from the [command line](https://docs.godotengine.org/en/stable/tutorials/editor/command_line_tutorial.html) with the `--verbose` command line argument, do you notice shader cache lines being slowly printed while the editor opens?

**Comment by crestofthebeast** [comment]
Here's a (super scuffed, sorry) demonstration of the project being run with `--verbose`.

https://www.youtube.com/watch?v=WHFWTj5IWao

> * Related to [Display text on the splash screen on initial startup to indicate that shaders will be cached godot-proposals#3657](https://github.com/godotengine/godot-proposals/issues/3657).
> 
> 
> If you start the editor from the [command line](https://docs.godotengine.org/en/stable/tutorials/editor/command_line_tutorial.html) with the `--verbose` command line argument, do you notice shader cache lines being slowly printed while the editor opens?
> 
> You could also try [using a C++ profiler](https://docs.godotengine.org/en/latest/development/cpp/using_cpp_profilers.html) on a [debug build of the engine](https://github.com/Calinou/godot-debug-builds/releases) to see which functions are taking the most time to run. (Only 4.0 binaries are available so far, but I'll add 4.0.1/4.0.2/4.0.3 in due time.)

**Comment by crestofthebeast** [comment]
Update -- 4.1 beta 1 seems to have noticeably faster startup times?
Would like to know if anyone else has experienced this.

**Comment by hudwave** [comment]
Yeah this seems to be fixed/working as intended for me on 4.1 beta 2

**Comment by Calinou** [comment]
Can anyone try this workaround from https://github.com/godotengine/godot/issues/78369?

> I had Forward+ renderer as default. Strangely, if I switch renderer to Mobile / Compatible, and then back to Forward+ (which causes editor to restart), the delay disappears. If I close and open Godot manually, delay will be there again.

**Comment by alexamy** [comment]
Can anyone try 4.1? Should be fixed.

**Comment by kaissouDev** [comment]
> Can anyone try 4.1? Should be fixed.

I've tried it and found that the start-up time is faster.

**Comment by akien-mga** [comment]
Thanks everyone for testing, closing as fixed in 4.1.
